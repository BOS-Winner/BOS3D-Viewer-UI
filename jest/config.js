module.exports = {
  rootDir: '../',
  testMatch: [
    "**/__tests__/**/*.test.{js,jsx}",
  ],
  moduleNameMapper: {
    "\\.less$": "identity-obj-proxy",
    "\\.css$": "<rootDir>/jest/__mocks__/styleMock.js",
    "Libs(.*)$": "<rootDir>/src/Libs$1",
    "UIutils(.*)$": "<rootDir>/src/UI/utils$1",
    "^Base(.*)$": "<rootDir>/src/UI/Base$1",
    "^BOS3D(.*)$": "<rootDir>/jest/__mocks__/BOS3D$1",
    "^BOS2D$": "<rootDir>/jest/__mocks__/BOS2D.js",
    "^BOS3DUI$": "<rootDir>/jest/__mocks__/BOS3DUI.js",
    "^jest/(.*)$": "<rootDir>/jest/$1",
    "^mock/(.*)$": "<rootDir>/jest/__mocks__/$1",
  },
  moduleFileExtensions: ["js", "jsx"],
  moduleDirectories: ["node_modules"],
  transform: {
    "\\.(js|jsx)": "babel-jest",
    "\\.(jpg|jpeg|png|gif|svg)$": "<rootDir>/jest/__mocks__/fileMock.js",
  },
  transformIgnorePatterns: [
    // Change MODULE_NAME_HERE to your module that isn't being compiled
    // "/node_modules/(?!three|lodash-es).+\\.js$"
    "/node_modules/core-js",
    "/node_modules/@testing-library",
    "/node_modules/@babel",
  ],
  coveragePathIgnorePatterns: [
    "/node_modules/",
    "/jest/"
  ],
};
