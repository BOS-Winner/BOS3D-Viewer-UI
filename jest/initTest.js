import EventEmitter from "events";
import { createStore } from "redux";
import BOS3D from 'BOS3D';
import reducer from "../src/UI/reducer";

export function genStore() {
  const ee = new EventEmitter();
  const viewer3D = new BOS3D.Viewer();

  return createStore(reducer, {
    system: {
      viewer3D,
      BIMWINNER: { BOS3D },
      apiVersion: 'api',
      eventEmitter: ee,
    }
  });
}
