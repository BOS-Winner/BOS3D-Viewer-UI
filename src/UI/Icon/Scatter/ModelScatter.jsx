import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
// import Icon from "Base/Icon";
// // import scatterPng from "../img/white/modelScatter.png";
// import style from "./style.less";
// import { AntdIcon } from '../../utils/utils';
// import iconstyle from '../../Toolbar/bottom.less';

/**
 * 模型分解
 */
class ModelScatter extends React.Component {
  static propTypes = {
    viewer: PropTypes.object.isRequired,
    enable: PropTypes.bool.isRequired,
    coefficient: PropTypes.number.isRequired,
  };

  shouldComponentUpdate(nextProps) {
    if (nextProps.enable && nextProps.coefficient !== this.props.coefficient) {
      this.onRangeChange(nextProps.coefficient);
      return false;
    }
    return true;
  }

  componentDidUpdate(prevProps) {
    if (prevProps.enable && !this.props.enable) {
      const viewer3D = this.props.viewer;
      const modelKeys = viewer3D.getViewerImpl().modelManager.getModelKeys();
      viewer3D.closeModelsExplosion(modelKeys);
    }
  }

  onRangeChange(value) {
    const coefficient = parseFloat(value);
    const viewer3D = this.props.viewer;
    const modelKeys = viewer3D.getViewerImpl().modelManager.getModelKeys();
    viewer3D.modelsExplosion({
      modelKey: modelKeys,
      coefficientX: coefficient,
      coefficientY: coefficient,
      coefficientZ: coefficient,
    });
  }

  render() {
    return (
      <div
        // className={style.rangeContainer}
        title="模型分解"
      >

        {/* <Icon
          icon={<AntdIcon type="icondecompose" className={iconstyle.icon} />}
          title="模型"
          showTitle={false}
        /> */}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  viewer: state.system.viewer3D,
});
const mapDispatchToProps = () => ({});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ModelScatter);
