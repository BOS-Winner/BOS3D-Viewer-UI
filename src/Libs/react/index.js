import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
