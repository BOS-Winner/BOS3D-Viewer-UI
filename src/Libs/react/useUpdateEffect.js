import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
