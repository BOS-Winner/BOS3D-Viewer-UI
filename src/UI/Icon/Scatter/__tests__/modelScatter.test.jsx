import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { genStore } from 'jest/initTest';
import ModelScatter from '../ModelScatter';

const store = genStore();
store.getState().system.viewer3D.modelsExplosion = jest.fn();
store.getState().system.viewer3D.closeModelsExplosion = jest.fn();
let scatterContainer;
let scatterRerender;

describe('model scatter', () => {
  beforeEach(() => {
    const { container, rerender } = render((
      <Provider store={store}>
        <ModelScatter enable coefficient={1} />
      </Provider>
    ));
    scatterContainer = container;
    scatterRerender = rerender;
  });

  it('should render dom', () => {
    expect(scatterContainer).toMatchSnapshot();
  });

  it('should scatter model', () => {
    const viewer3D = store.getState().system.viewer3D;
    scatterRerender((
      <Provider store={store}>
        <ModelScatter enable coefficient={1.5} />
      </Provider>
    ));
    expect(viewer3D.modelsExplosion).toBeCalledTimes(1);
    expect(viewer3D.modelsExplosion.mock.calls[0]).toMatchSnapshot();

    scatterRerender((
      <Provider store={store}>
        <ModelScatter enable={false} coefficient={2} />
      </Provider>
    ));
    expect(viewer3D.modelsExplosion).toBeCalledTimes(1);
    expect(viewer3D.closeModelsExplosion).toBeCalledTimes(1);
    expect(viewer3D.closeModelsExplosion.mock.calls[0]).toMatchSnapshot();
  });
});
