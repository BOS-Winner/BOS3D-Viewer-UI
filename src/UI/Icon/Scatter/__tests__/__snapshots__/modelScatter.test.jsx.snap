// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model scatter should render dom 1`] = `
<div>
  <div
    class="rangeContainer"
    title="模型分解"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(modelScatter.png);"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`model scatter should scatter model 1`] = `
Array [
  Object {
    "coefficientX": 1.5,
    "coefficientY": 1.5,
    "coefficientZ": 1.5,
    "modelKey": Array [
      "M1145141919810",
      "M1145141919811",
    ],
  },
]
`;

exports[`model scatter should scatter model 2`] = `
Array [
  Array [
    "M1145141919810",
    "M1145141919811",
  ],
]
`;
